homeassistant/components/satel_integra/__init__.py

```diff
@@ -93,61 +89,86 @@ def is_alarm_code_necessary(value):
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the Satel Integra component."""
-    conf = config[DOMAIN]
 
-    zones = conf.get(CONF_ZONES)
-    outputs = conf.get(CONF_OUTPUTS)
-    switchable_outputs = conf.get(CONF_SWITCHABLE_OUTPUTS)
-    host = conf.get(CONF_HOST)
-    port = conf.get(CONF_PORT)
-    partitions = conf.get(CONF_DEVICE_PARTITIONS)
+    conf = config.get(DOMAIN)
+
+    if conf is None:
+        # If we have a config entry, setup is done by that config entry.
+        # If there is no config entry, this should fail.
+        return bool(hass.config_entries.async_entries(DOMAIN))
+
+    conf = dict(conf)
+    hass.data[DATA_SATEL_CONFIG] = conf
+
+    # Only import if we haven't before.
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data=conf
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Load a config entry."""
+    if (conf := hass.data.get(DATA_SATEL_CONFIG)) is None:
+        _conf = await async_integration_yaml_config(hass, DOMAIN)
+        if not _conf or DOMAIN not in _conf:
+            _LOGGER.warning(
+                "No `satel:` key found in configuration.yaml. See "
+                "https://www.home-assistant.io/integrations/satel_integra/ "
+                "for satel entity configuration documentation"
+            )
+            # generate defaults
+            conf = CONFIG_SCHEMA({DOMAIN: {}})[DOMAIN]
+        else:
+            conf = _conf[DOMAIN]
+
+    zones = conf[CONF_ZONES]
+    outputs = conf[CONF_OUTPUTS]
+    switchable_outputs = conf[CONF_SWITCHABLE_OUTPUTS]
+    partitions = conf[CONF_DEVICE_PARTITIONS]
+
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
 
     monitored_outputs = collections.OrderedDict(
         list(outputs.items()) + list(switchable_outputs.items())
     )
 
-    controller = AsyncSatel(host, port, hass.loop, zones, monitored_outputs, partitions)
+    try:
+        controller = AsyncSatel(
+            host,
+            port,
+            hass.loop,
+            zones,
+            monitored_outputs,
+            partitions,
+        )
+
+        result = await controller.connect()
 
-    hass.data[DATA_SATEL] = controller
+        if not result:
+            raise Exception("Controller failed to connect")
 
-    result = await controller.connect()
+    except Exception as ex:
+        raise ConfigEntryNotReady from ex
 
-    if not result:
-        return False
+    hass.data[DATA_SATEL_CONFIG] = conf
+    hass.data[DOMAIN] = controller
 
     @callback
     def _close(*_):
         controller.close()
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _close)
+    entry.async_on_unload(hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _close))
```

Comment:
Why do we need to wait for home assistant to stop before closing the connection? Can't it be closed right away on entry unload?